composer/distributed/fsdp2.py

```diff
@@ -4,15 +4,77 @@
 """Helpers for FSDP2."""
 
 import warnings
-from typing import Optional
+from typing import Any, Callable, Optional, Union
 
 import torch
 import torch.nn as nn
 from torch.distributed.fsdp._fully_shard import fully_shard
+from torch.distributed.fsdp.wrap import CustomPolicy
 
 from composer.utils.parallelism import FSDP2Config
 
 
+def _generate_default_policy(parent_model: nn.Module) -> CustomPolicy:
+    # The same policy as FSDP1 with some caveats around the parent_model (root_module)
+    def lambda_fn(current_module: nn.Module) -> Union[bool, dict[str, Any]]:
+        ret = False
+        if hasattr(current_module, '_fsdp_wrap'):
+            ret = bool(current_module._fsdp_wrap)
+        elif hasattr(parent_model, 'fsdp_wrap_fn') and isinstance(parent_model.fsdp_wrap_fn, Callable):
+            # There are certain situations where _fsdp_wrap for the parent model is not set, but we wrap submodules
+            # with _fsdp_wrap_fn (e.g. wrapping all GPTBlocks). In those situations, we generally also want to wrap
+            # the parent model, so we have an additional check here.
+            if current_module == parent_model:
+                return True
+            ret = parent_model.fsdp_wrap_fn(current_module)
+            if isinstance(ret, dict):
+                assert set(ret.keys()) == {'mesh', 'reshard_after_forward'} \
```

Comment:
Can we make this so you don't have to specify all args? Like it should inherit the default config and this only applies anything in `ret` as an override

Reply:
also seems like we still need to add a test for using fsdp_wrap_fn and returning kwargs?